src/main/java/org/folio/services/validator/engine/ValidationEngineServiceImpl.java

```diff
@@ -0,0 +1,276 @@
+            errorMessagesFuture.setHandler(asyncResult -> {
+              if (asyncResult.succeeded()) {
+                prepareResponse(asyncResult.result(), resultHandler);
+              } else {
+                resultHandler.handle(Future.failedFuture(asyncResult.cause().getMessage()));
+              }
+            });
+          }
+        });
+      } else {
+        resultHandler.handle(Future.failedFuture(rulesResponse.cause().getMessage()));
+      }
+    });
+  }
+
+  private List<Rule> prepareRulesBeforeValidation(List<Rule> rules, AsyncResult<JsonObject> lookupUserHandler) {
+    JsonObject user = lookupUserHandler.result();
+    String userName = user.getString("username");
+    for (Rule rule : rules) {
+      if (Rule.Type.REG_EXP.equals(rule.getType())) {
+        rule.setExpression(rule.getExpression().replace(REGEXP_USER_NAME_PLACEHOLDER, userName));
+      }
+    }
+    return rules.stream().sorted(Comparator.comparing(Rule::getOrderNo)).collect(Collectors.toList());
+  }
+
+  private Future<List<String>> validatePasswordByRules(final List<Rule> rules,
+                                                       final String password,
+                                                       final MultiMap headers) {
+    List<String> errorMessages = new ArrayList(rules.size());
+    rules.sort(Comparator.comparing(Rule::getOrderNo));
+
+    Future<List<String>> future = Future.future();
+    List<Future> programmaticRulesFutures = new ArrayList<>();
+    for (Rule rule : rules) {
+      if (Rule.Type.REG_EXP.equals(rule.getType())) {
+        validatePasswordByRexExpRule(password, rule, errorMessages);
+      } else if (Rule.Type.PROGRAMMATIC.equals(rule.getType())) {
+        programmaticRulesFutures.add(getValidatePasswordByProgrammaticRuleFuture(password, rule, errorMessages, headers));
+      }
+    }
+    // Notify external method future handler when all programmatic rule futures complete
+    CompositeFuture.all(programmaticRulesFutures).setHandler(compositeFutureAsyncResult -> {
+      if (compositeFutureAsyncResult.succeeded()) {
+        future.complete(errorMessages);
+      } else {
+        future.fail(compositeFutureAsyncResult.cause().getMessage());
+      }
+    });
+    return future;
+  }
+
+  private void validatePasswordByRexExpRule(final String password,
+                                            final Rule rule,
+                                            final List<String> errorMessages) {
+    String expression = rule.getExpression();
+    if (!Pattern.compile(expression).matcher(password).matches()) {
+      errorMessages.add(rule.getErrMessageId());
+    }
+  }
+
+  private Future<JsonObject> lookupUser(MultiMap headers) {
+    Future<JsonObject> future = Future.future();
+    String userId = headers.get(OKAPI_USERID_HEADER);
+    String okapiUrl = headers.get(OKAPI_URL_HEADER);
+    String userNameRequestUrl = String.format("%s/users?query=id==%s", okapiUrl, userId);
```

Comment:
Using the single record endpoint is more easy to use and likely to be faster:
https://s3.amazonaws.com/foliodocs/api/mod-users/users.html
String.format("%s/users/%s", okapiUrl, userId)